File: Tree/BinaryTree.cpp
#include <iostream>
#include <climits>
using namespace std;

// 结点
struct Node {
    int data;
    Node* lchild;
    Node* rchild;

    Node() {
        data = INT_MAX;
        lchild = nullptr;
        rchild = nullptr;
    }
};

// 二叉树
class BinaryTree {
private:

    // -----------
    // Attributes
    // -----------
    
    Node* root; 
    int size;   

    // 前序遍历
    void preOrderTraverse(Node* node) {
        if (root != nullptr) {
            cout << root->data << " ";
            preOrderTraverse(root->lchild);
            preOrderTraverse(root->rchild);
        }
    }

    // 中序遍历
    void inOrderTraverse(Node* node) {
        if (root != nullptr) {
            inOrderTraverse(root->lchild);
            cout << root->data << " ";
            inOrderTraverse(root->rchild);
        }
    }

    // 后序遍历
    void postOrderTraverse(Node* node) {
        if (root != nullptr) {
            postOrderTraverse(root->lchild);
            postOrderTraverse(root->rchild);
            cout << root->data << " ";
        }
    }

    // 插入结点
    Node* insertData(Node* node, int data) {
        // 根节点为空，则说明树不存在，先建立根节点
        if (root == nullptr) {
            root->data = data;
            setSize(getSize() + 1);
        }
        else {
            Node* newNode = new (Node);
            newNode->data = data;
            if (data < root->data) {
                root->lchild = insertData(root->lchild, data);
            }
            else if (data > root->data) {
                root->rchild = insertData(root->rchild, data);
            }
            setSize(getSize() + 1);
        }    
        return root;
    }

public:
    BinaryTree() {
        root = nullptr;
        size = 0;
    }

    ~BinaryTree() {}

    Node* getRoot() {
        return root;
    }

    int getSize() {
        return size;
    }

    void setSize(int size) {
        this->size = size;
    }

    // 前序遍历API
    void preOrder() {
        this->preOrderTraverse(this->getRoot());
    }

    // 中序遍历API
    void inOrder() {
        this->inOrderTraverse(this->getRoot());
    }

    // 后序遍历API
    void postOrder() {
        this->postOrderTraverse(this->getRoot());
    }

    // 插入结点API
    void insert(int data) {
        this->insertData(this->getRoot(), data);
    }

};

int main() {
    BinaryTree myTree = BinaryTree();
    int a[11] = {50, 30, 10, 0, 20, 40, 70, 90, 100, 60, 80};
    for (int i = 0; i < 11; i++) {
        myTree.insert(a[i]);
    } 
    cout << "前序遍历" << endl;
    myTree.preOrder();
    return 0;
}

File: Tree/python/BinaryTree.py
class Node:
    def __init__(self, value, left=None, right=None) -> None:
        self.value = value
        self.left = left
        self.right =right

class BinaryTree:
    def __init__(self, root=None) -> None:
        self.root = root
    
    # 添加结点
    def add(self, value):
        node = Node(value)
        if self.root is None:
            self.root = node
            return

        # 通过队列来找到待添加的结点位置
        queue = [self.root]
        curr_node = queue.pop(0)
        # 插入到左子树
        if curr_node.left is None:
            curr_node.left = node
            return
        # 将左孩子插入到队列中，下一次开始循环左结点
        else:
            queue.append(curr_node.left)
        
        if curr_node.right is None:
            curr_node.right = node
            return 
        else:
            queue.append(curr_node.right)
    
    # 层序遍历（广度优先遍历思想）
    def breadth_traverse(self):
        if self.root is None:
            return 

        # 通过维护一个队列来一层一层地进行扫描
        queue = [self.root]
        while queue:
            curr_node = queue.pop(0)
            print(curr_node.value)

            if curr_node.left is not None:
                queue.append(curr_node.left)
            if curr_node.right is not None:
                queue.append(curr_node.right)
    
    # 前序遍历
    def pre_order_traverse(self):
        if not self.root:
            return
        print(self.root.value)
        self.pre_order_traverse(self.root.left)
        self.pre_order_traverse(self.root.right)

    # 中序遍历
    def in_order_traverse(self):
        if not self.root:
            return 
        self.in_order_traverse(self.root.left)
        print(self.root.value)
        self.in_order_traverse(self.root.right)

    # 后序遍历
    def post_order_traverse(self):
        if not self.root:
            return
        self.post_order_traverse(self.root.left)
        self.post_order_traverse(self.root.right)
        print(self.root.value)